### PR TITLE
System comments

### LINKED ISSUE #746: Show system comments for bugs [30h]
> Follow-up for #685.

System events should be shown along with bug comments.
System comment should not be editable or removable.

Log at least these system events
- This bug was also found in ...
- Bug was suppressed ...
- Bug detection status changed in run run1 to resolved
- Bug review status changed from .. to ..

Implementation

- Add a kind field to the Bug api/report_server.thrift
- Store system events text in the code only

In the database store system comments as formatted strings for example:
"rev_st_changed Unresolved Resolved"
Make sure escaping spaces is done. Maybe use shlex or some other magic.

Create a lookup table in the config files:
```json
{
"rev_st_changed": "Bug review status changed from %s to %s"
} 
```

Make sure the "token" is not too long, so the database doesn't store too much bloat.

Which would print to the user:
- Bug review status changed from Unresolved to Resolved.

Update GUI: 4h
Extend API : 2h
Implement formatting: 8h
Introduce system events: 8h
Test cases: 8h

### PR DESCRIPTION
**Pull request by csordasmarton**
> Closes #746

Show system comments at bugs.

Screenshot:
image

### CODE REVIEW

**Comment by whisperity** [comment]
It'd be even fancier if the "blips" could also be turned into icons (like how bug path elements have an icon, or how the blips of moments on GitHub on this patch also have icons), but that's just continuing to polish it fancy...

I really like the whole "bus schedule" looks to it. :heart_eyes:

File: web/api/v6/report_server.thrift [hunk 1]
```diff
@@ -85,6 +85,11 @@ enum ExtendedReportDataType {
   FIXIT = 20,
 }
 
+enum CommentKind {
+  GENERAL, // User-given comments.
```
Comment (gyorb) [inline web/api/v6/report_server.thrift]: I would rename this to USER, GENERAL is a bit too general ;)

File: web/server/codechecker_server/api/report_server.py [hunk 2]
```diff
@@ -313,6 +314,20 @@ def extended_data_db_to_api(erd):
         fileId=erd.file_id)
 
 
+def comment_kind_str(status):
+    if status == ttypes.CommentKind.GENERAL:
+        return 'general'
```
Comment (gyorb) [inline web/server/codechecker_server/api/report_server.py]: The same here `user`

File: web/config/system_comment_kinds.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
+  "rev_st_changed": "%author% changed review status from <b class=\"old-review-status\">{0}</b> to <b class=\"new-review-status\">{1}</b> %date%",
```
Comment (gyorb) [inline web/config/system_comment_kinds.json]: Do we need to store the whole text, couldn't we create some more dense representation?

Comment (csordasmarton) [reply web/config/system_comment_kinds.json]: We are not store the whole text in the database we just store the type of the system comment and the arguments: `rev_st_changed Unreviewed Confirmed`. We are using this file only to display the comment in the gui.

File: web/server/codechecker_server/api/report_server.py [hunk 4]
```diff
@@ -1186,12 +1236,26 @@ def getComments(self, report_id):
                     .order_by(Comment.created_at.desc()) \
                     .all()
 
+                context = webserver_context.get_context()
                 for comment in comments:
+                    message = comment.message
+
+                    if comment.kind == 'system':
```
Comment (gyorb) [inline web/server/codechecker_server/api/report_server.py]: Please convert the Enum value to string for the comparison so it is harder to make mistakes during a renaming.

```suggestion
                    if comment.kind == comment_kind_str(ttypes.CommentKind.SYSTEM):
```

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 4
           

Complexity increasing per file
==============================
- web/server/codechecker_server/migrations/report/versions/6cb6a3a41967_system_comments.py  1
         

Clones added
============
- web/tests/functional/comment/test_comment.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/codechecker/codechecker/pullRequest?prid=3374942&bid=12152105)

File: web/server/www/scripts/codecheckerviewer/CommentView.js [hunk 5]
```diff
@@ -121,31 +120,25 @@ function (declare, dom, style, topic, Memory, Observable, ConfirmDialog,
...
           var newContent = that._commentContent.get('value');
 
           try {
-            CC_SERVICE.updateComment(that.cId, newContent);
+            CC_SERVICE.updateComment(that.commentData.id, newContent);
           } catch (ex) { util.handleThriftException(ex); }
 
-          that._message.set('content',
-            newContent.replace(/(?:\r\n|\r|\n)/g, '<br>'));
+          that._message.innerHTML =
```
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: [Unsafe assignment to innerHTML (no-unsafe-innerhtml/no-unsafe-innerhtml)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/server/www/scripts/codecheckerviewer/CommentView.js [hunk 6]
```diff
@@ -158,6 +151,54 @@ function (declare, dom, style, topic, Memory, Observable, ConfirmDialog,
...
+    },
+
+    addStatusIcons : function (domElement) {
+      ['detection-status', 'review-status'].forEach(function (type) {
+        ['old-' + type, 'new-' + type].forEach(function (className) {
+          var domStatuses =
+            domElement.getElementsByClassName(className);
+
+          for (var i = 0; i < domStatuses.length; i++) {
+            var domStatus = domStatuses[i];
```
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: [Variable Assigned to Object Injection Sink (security/detect-object-injection)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/server/www/scripts/codecheckerviewer/CommentView.js [hunk 7]
```diff
@@ -108,7 +107,7 @@ function (declare, dom, style, topic, Memory, Observable, ConfirmDialog,
         content   : 'Are you sure you want to delete this?',
         onExecute : function () {
           try {
-            CC_SERVICE.removeComment(that.cId);
+            CC_SERVICE.removeComment(that.commentData.id);
```
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: ['CC_SERVICE' is not defined. (no-undef)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/server/www/scripts/codecheckerviewer/CommentView.js [hunk 8]
```diff
@@ -209,15 +250,19 @@ function (declare, dom, style, topic, Memory, Observable, ConfirmDialog,
...
-          that._comments.addChild(new Comment({
-            class    : 'comment',
-            reportId : reportId,
-            cId      : comment.id,
-            author   : comment.author,
-            time     : comment.createdAt,
-            message  : comment.message,
-            sender   : sender
-          }));
+          if (comment.kind === CC_OBJECTS.CommentKind.SYSTEM) {
```
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: ['CC_OBJECTS' is not defined. (no-undef)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 4
           

Complexity increasing per file
==============================
- web/server/codechecker_server/migrations/report/versions/6cb6a3a41967_system_comments.py  1
         

Clones added
============
- web/tests/functional/comment/test_comment.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942&bid=12152105)

File: web/server/www/scripts/codecheckerviewer/CommentView.js (hunk 5 above)
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: [Unsafe assignment to innerHTML (no-unsafe-innerhtml/no-unsafe-innerhtml)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/server/www/scripts/codecheckerviewer/CommentView.js (hunk 6 above)
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: [Variable Assigned to Object Injection Sink (security/detect-object-injection)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/server/www/scripts/codecheckerviewer/CommentView.js (hunk 7 above)
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: ['CC_SERVICE' is not defined. (no-undef)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/server/www/scripts/codecheckerviewer/CommentView.js (hunk 8 above)
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: ['CC_OBJECTS' is not defined. (no-undef)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 4
           

Complexity increasing per file
==============================
- web/server/codechecker_server/migrations/report/versions/6cb6a3a41967_system_comments.py  1
         

Clones added
============
- web/tests/functional/comment/test_comment.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942&bid=12152105)

File: web/server/www/scripts/codecheckerviewer/CommentView.js (hunk 5 above)
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: [Unsafe assignment to innerHTML (no-unsafe-innerhtml/no-unsafe-innerhtml)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/server/www/scripts/codecheckerviewer/CommentView.js (hunk 6 above)
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: [Variable Assigned to Object Injection Sink (security/detect-object-injection)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/server/www/scripts/codecheckerviewer/CommentView.js (hunk 7 above)
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: ['CC_SERVICE' is not defined. (no-undef)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/server/www/scripts/codecheckerviewer/CommentView.js (hunk 8 above)
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: ['CC_OBJECTS' is not defined. (no-undef)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/server/codechecker_server/migrations/report/versions/6cb6a3a41967_system_comments.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+down_revision = '3e91d0612422'
+branch_labels = None
+depends_on = None
+
+from alembic import op
+import sqlalchemy as sa
+
+
+def upgrade():
+    op.add_column('comments', sa.Column('kind', sa.Enum(u'user',
```
Comment (csordasmarton) [inline web/server/codechecker_server/migrations/report/versions/6cb6a3a41967_system_comments.py]: It will fail on postgresql. We can fix this by using the following instructions: https://stackoverflow.com/questions/37848815/sqlalchemy-postgresql-enum-does-not-create-type-on-db-migrate.
```py
comment_kind_name = 'comment_kind'
enum_values = (u'user', u'system')

commnet_kind_column = sa.Column('kind', sa.Enum(*enum_values,
                                                name=comment_kind_name),
                                server_default=u'user',
                                nullable=False)
def upgrade():
    ctx = op.get_context()
    dialect = ctx.dialect.name

    if dialect == 'postgresql':
        from sqlalchemy.dialects import postgresql
        comment_kind = postgresql.ENUM(*enum_values, name=comment_kind_name)
        comment_kind.create(op.get_bind())

        op.add_column('comments', commnet_kind_column)
    elif dialect == 'sqlite':
        op.add_column('comments', commnet_kind_column)
def downgrade():
    op.drop_column('comments', 'kind')

    ctx = op.get_context()
    dialect = ctx.dialect.name

    if dialect == 'postgresql':
        from sqlalchemy.dialects import postgresql
        comment_kind = postgresql.ENUM(*enum_values, name=comment_kind_name)
        comment_kind.drop(op.get_bind())
```

But as we discussed it personally we will use integers instead of enums to avoid this kind of hacks.

**Comment by csordasmarton** [comment]
@gyorb I have refactored the migration script to use Integer instead of Enum.

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 4
           

Complexity increasing per file
==============================
- web/server/codechecker_server/migrations/report/versions/6cb6a3a41967_system_comments.py  1
         

Clones added
============
- web/tests/functional/comment/test_comment.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942&bid=12152105)

File: web/server/www/scripts/codecheckerviewer/CommentView.js (hunk 5 above)
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: [Unsafe assignment to innerHTML (no-unsafe-innerhtml/no-unsafe-innerhtml)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/server/www/scripts/codecheckerviewer/CommentView.js (hunk 6 above)
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: [Variable Assigned to Object Injection Sink (security/detect-object-injection)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/server/www/scripts/codecheckerviewer/CommentView.js (hunk 7 above)
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: ['CC_SERVICE' is not defined. (no-undef)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/server/www/scripts/codecheckerviewer/CommentView.js (hunk 8 above)
Comment (gyorb) [inline web/server/www/scripts/codecheckerviewer/CommentView.js]: Codacy Issue found: ['CC_OBJECTS' is not defined. (no-undef)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942)

File: web/api/v6/report_server.thrift [hunk 10]
```diff
@@ -93,6 +93,11 @@ enum ExtendedReportDataType {
   FIXIT = 20,
 }
 
+enum CommentKind {
```
Comment (gyorb) [inline web/api/v6/report_server.thrift]: I'm thinking maybe it would be better if we introduce some prefix for the thrift types.
There are conversion functions which convert from CommentKind to CommentKind one of them is a thrift enum and the other is a local enum it is hard to follow which type is used where and what is converted to what.

File: web/server/codechecker_server/api/report_server.py [hunk 11]
```diff
@@ -469,6 +477,20 @@ def extended_data_db_to_api(erd):
         fileId=erd.file_id)
 
 
+def comment_kind_str(status):
```
Comment (gyorb) [inline web/server/codechecker_server/api/report_server.py]: Is there any string involved in this conversion? Just another enum value is returned.
I'm thinking would it be better to tie this conversion functions closer to the CommentKind local enum type?
With some other name which makes the conversion more explicit to_thrift_type, from_thrift_type.

A function docstring would be nice for the conversion functions.

File: web/server/codechecker_server/api/report_server.py [hunk 12]
```diff
@@ -653,6 +675,17 @@ def sort_run_data_query(query, sort_mode):
     return query
 
 
+def escape_whitespaces(s, whitespaces=None):
```
Comment (gyorb) [inline web/server/codechecker_server/api/report_server.py]: Please add function docstring.

File: web/server/codechecker_server/api/report_server.py [hunk 13]
```diff
@@ -1397,12 +1455,27 @@ def getComments(self, report_id):
                     .order_by(Comment.created_at.desc()) \
                     .all()
 
+                context = webserver_context.get_context()
                 for comment in comments:
+                    message = comment.message
+
+                    sys_comment = comment_kind_str(ttypes.CommentKind.SYSTEM)
+                    if comment.kind == sys_comment:
+                        elements = shlex.split(message)
```
Comment (gyorb) [inline web/server/codechecker_server/api/report_server.py]: why do we need the shlex.split here?

File: web/server/codechecker_server/api/report_server.py [hunk 14]
```diff
@@ -2441,6 +2520,7 @@ def checker_is_unavailable(checker_name):
                 detection_status = 'new'
                 detected_at = run_history_time
 
+                old_status = None
```
Comment (gyorb) [inline web/server/codechecker_server/api/report_server.py]: Do we need to set the default value here? The variable is used only inside the if block below I did not found any other references.

File: web/server/codechecker_server/database/run_db_model.py [hunk 15]
```diff
@@ -360,12 +360,16 @@ class Comment(Base):
     bug_hash = Column(String, nullable=False, index=True)
     author = Column(String, nullable=False)
     message = Column(Binary, nullable=False)
+    kind = Column(Integer,
+                  nullable=False,
+                  server_default="0")
```
Comment (gyorb) [inline web/server/codechecker_server/database/run_db_model.py]: Maybe a comment about the default value here would be nice. '0' means a comment type by the user tight?

File: web/server/codechecker_server/api/report_server.py (hunk 13 above)
Comment (csordasmarton) [reply web/server/codechecker_server/api/report_server.py]: Before we store a system comment in the database, we will escape the white spaces (`escape_whitespaces`). We are using `shlex` here to split this escaped string properly.

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 4
           

Complexity increasing per file
==============================
- web/server/codechecker_server/migrations/report/versions/6cb6a3a41967_system_comments.py  1
         

Clones added
============
- web/tests/functional/comment/test_comment.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942&bid=12152105)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 6
           

Complexity increasing per file
==============================
- web/server/codechecker_server/migrations/report/versions/6cb6a3a41967_system_comments.py  1
- web/tests/functional/comment/test_comment.py  2
         

Clones added
============
- web/tests/functional/comment/test_comment.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942&bid=12152105)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 6
           

Complexity increasing per file
==============================
- web/server/codechecker_server/migrations/report/versions/6cb6a3a41967_system_comments.py  1
- web/tests/functional/comment/test_comment.py  2
         

Clones added
============
- web/tests/functional/comment/test_comment.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=3374942&bid=12152105)

**Comment by whisperity** [comment]
Oho, this is amazing, thank you!
